sim/core/spell_result.go

```diff
@@ -292,16 +287,40 @@ func (spell *Spell) calcDamageInternal(sim *Simulation, target *Unit, baseDamage
 }
 func (spell *Spell) CalcDamage(sim *Simulation, target *Unit, baseDamage float64, outcomeApplier OutcomeApplier) *SpellResult {
 	attackerMultiplier := spell.AttackerDamageMultiplier(spell.Unit.AttackTables[target.UnitIndex][spell.CastType])
+	if spell.BonusCoefficient > 0 {
+		baseDamage += spell.BonusCoefficient * spell.SpellDamage()
+	}
 	return spell.calcDamageInternal(sim, target, baseDamage, attackerMultiplier, false, outcomeApplier)
 }
 func (spell *Spell) CalcPeriodicDamage(sim *Simulation, target *Unit, baseDamage float64, outcomeApplier OutcomeApplier) *SpellResult {
 	attackerMultiplier := spell.AttackerDamageMultiplier(spell.Unit.AttackTables[target.UnitIndex][spell.CastType])
+	if spell.Dot(target).BonusCoefficient > 0 {
+		baseDamage += spell.Dot(target).BonusCoefficient * spell.SpellDamage()
+	}
 	return spell.calcDamageInternal(sim, target, baseDamage, attackerMultiplier, true, outcomeApplier)
 }
 func (dot *Dot) CalcSnapshotDamage(sim *Simulation, target *Unit, outcomeApplier OutcomeApplier) *SpellResult {
 	return dot.Spell.calcDamageInternal(sim, target, dot.SnapshotBaseDamage, dot.SnapshotAttackerMultiplier, true, outcomeApplier)
 }
 
+func (dot *Dot) Snapshot(target *Unit, baseDamage float64, isRollover bool) {
+	dot.SnapshotBaseDamage = baseDamage
+	if dot.BonusCoefficient > 0 {
+		dot.SnapshotBaseDamage += dot.BonusCoefficient * dot.Spell.SpellDamage()
+	}
+
+	if !isRollover {
+		attackTable := dot.Spell.Unit.AttackTables[target.UnitIndex][dot.Spell.CastType]
+		dot.SnapshotAttackerMultiplier = dot.Spell.AttackerDamageMultiplier(attackTable)
+	}
```

Comment:
Rollovers in SOD dont update anything except the duration. Even SP is snapshot from the original cast

Reply:
Yeah, that could also be done and the `OnShapshot` can instead be called only on application and not on rollover from the dot code

Reply:
Tho when I think about it again blizzard can decide to make rollovers recalculate sp again and then we will have to revert it back. I think its better to leave the rollover support in the snapshot function and fix it to be properly passed to the shortcut

Reply:
Changed the default rollover behavior. 

Imho DoTs could be simplified by adding default behavior for things like OnApplication, OnRollover and OnStackChange. Especially if AP would be handled similar to how SP is now handled in this PR, there would be no reason to do anything custom for most spells. The separation of Spell and DoT is also not the best choice, because both have common data and logic when cast/applied. There should really be SpellEffects, pretty much like the game does it. But that's all far beyond this PR.

Reply:
I'm still voting for "Snapshot()" to support only the default case, and ignore rollovers or stacking DOTs. It's a very helpful utility function as it is, and it's easier to use if it just does one thing.
Rollovers simply wouldn't call it at all, stacking DOTs could use it for the initial application.